eth/transactionManager_test.go

```diff
@@ -0,0 +1,320 @@
+	gpm := &GasPriceMonitor{
+		minGasPrice: big.NewInt(0),
+		maxGasPrice: big.NewInt(0),
+		gasPrice:    big.NewInt(1),
+	}
+	tm := &TransactionManager{
+		cond:      sync.NewCond(&sync.Mutex{}),
+		eth:       eth,
+		queue:     q,
+		txTimeout: 2 * time.Second,
+		gpm:       gpm,
+	}
+
+	stubTx := types.NewTransaction(1, pm.RandAddress(), big.NewInt(100), 100000, gasPrice, pm.RandBytes(68))
+
+	// Test eth.TransactionByHash error
+	expErr := errors.New("TransactionByHash error")
+	eth.err["TransactionByHash"] = expErr
+
+	tx, err := tm.replace(stubTx)
+	assert.Nil(tx)
+	assert.EqualError(err, expErr.Error())
+	eth.err["TransactionByHash"] = nil
+
+	// Test no error - ErrReplacingMinedTx
+	eth.pending = false
+	tx, err = tm.replace(stubTx)
+	assert.Nil(tx)
+	assert.EqualError(err, ErrReplacingMinedTx.Error())
+
+	// Test error is eth.TransactionNotFound - fail at next step
+	eth.pending = true
+	gpm.maxGasPrice = big.NewInt(1)
+	eth.err["TransactionByHash"] = ethereum.NotFound
+	tx, err = tm.replace(stubTx)
+	assert.Nil(tx)
+	assert.EqualError(
+		err,
+		fmt.Sprintf("Replacement gas price exceeds max gas price suggested=%v max=%v", calcReplacementGasPrice(stubTx), gpm.maxGasPrice),
+	)
+	eth.err["TransactionByHash"] = nil
+
+	// Replacement gas price exceeds max gas price
+	// Throw error
+	tx, err = tm.replace(stubTx)
+	assert.Nil(tx)
+	assert.EqualError(
+		err,
+		fmt.Sprintf("Replacement gas price exceeds max gas price suggested=%v max=%v", calcReplacementGasPrice(stubTx), gpm.maxGasPrice),
+	)
+
+	// Error signing replacement tx
+	expErr = errors.New("SignTx error")
+	sig := &stubTransactionSigner{
+		err: nil,
+	}
+	tm.sig = sig
+	sig.err = expErr
+	gpm.maxGasPrice = big.NewInt(99999)
+	tx, err = tm.replace(stubTx)
+	assert.Nil(tx)
+	assert.EqualError(err, expErr.Error())
+	sig.err = nil
+
+	// Error sending replacement tx
+	expErr = errors.New("SendTx error")
+	eth.err["SendTransaction"] = expErr
+	logsBefore := glog.Stats.Info.Lines()
+	tx, err = tm.replace(stubTx)
+	logsAfter := glog.Stats.Info.Lines()
+	assert.EqualError(err, expErr.Error())
+	assert.Equal(logsAfter-logsBefore, int64(1))
+	eth.err["SendTransaction"] = nil
+
+	// Success
+	logsBefore = glog.Stats.Info.Lines()
+	tx, err = tm.replace(stubTx)
+	logsAfter = glog.Stats.Info.Lines()
+	assert.Nil(err)
+	expTx := types.NewTransaction(1, *stubTx.To(), stubTx.Value(), 100000, calcReplacementGasPrice(stubTx), stubTx.Data())
+	assert.Equal(tx, expTx)
+	assert.Equal(logsAfter-logsBefore, int64(1))
+
+	// Replacement gas price lower than suggest gas price
+	// Use market gas price
+	gpm.gasPrice = big.NewInt(999)
+	logsBefore = glog.Stats.Info.Lines()
+	tx, err = tm.replace(stubTx)
+	logsAfter = glog.Stats.Info.Lines()
+	assert.Nil(err)
+	expTx = types.NewTransaction(1, *stubTx.To(), stubTx.Value(), 100000, gpm.gasPrice, stubTx.Data())
+	assert.Equal(tx, expTx)
+	assert.Equal(logsAfter-logsBefore, int64(1))
+}
+
+func TestTransactionManager_CheckTxLoop(t *testing.T) {
```

Comment:
I think we should add a test to ensure that in the worst case we only submit maxReplacements number of replacement txs. This test would assert that maxReplacements number of replacement txs were submitted, but no more than that.

Reply:
Would something like this work 840c9f4 ?

Reply:
Looks good.